app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,236 @@
+      @analytics = Analytics.new(user: AnonymousUser.new, request: nil, session: {}, sp: nil)
+    end
+
+    # Checks the cache for an unexpired token and returns that.
+    # If the cache has expired, retrieves a new token and returns it
+    # @return [String] Auth token
+    def token
+      cache_value = Rails.cache.read(cache_key)
+      return cache_value unless IdentityConfig.store.arcgis_token_sync_request_enabled
+      if cache_value.nil?
+        retrieve_token&.fetch(:token)
+      else
+        handle_expired_token(cache_value)&.fetch(:token)
+      end
+    end
+
+    # @param [Object] cache_value the entry to save in cache
+    # @param [Number] expires_at the unix time cache entry should expire
+    def save_token(cache_value, expires_at)
+      Rails.cache.write(cache_key, cache_value, expires_at: expires_at)
+      # If using a redis cache we have to manually set the expires_at. This is because we aren't
+      # using a dedicated Redis cache and instead are just using our existing Redis server with
+      # mixed usage patterns. Without this cache entries don't expire.
+      # More at https://api.rubyonrails.org/classes/ActiveSupport/Cache/RedisCacheStore.html
+      Rails.cache.try(:redis)&.expireat(cache_key, expires_at.to_i)
+    end
+
+    # Makes a request to retrieve a new token
+    # it expires after 1 hour
+    # @return [String] Auth token
+    def retrieve_token
+      token, expires = request_token.fetch_values('token', 'expires')
+      expires_at = Time.zone.at(expires / 1000).to_f
+      cache_entry = {
+        token: token,
+        expires_at: expires_at,
+      }
+
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled == true
+        cache_entry = cache_entry.merge(
+          { sliding_expires_at: prefetch_ttl >= 0 ? expires_at - 3 * prefetch_ttl : expires_at },
+        )
+      end
+      save_token(cache_entry, expires_at)
+      cache_entry
+    end
+
+    def remove_token
+      Rails.cache.delete(cache_key)
+    end
+
+    private
+
+    def handle_expired_token(cache_value)
+      cache_value = wrap_raw_token(cache_value) unless cache_value.is_a?(Hash)
+      expires_at = cache_value.fetch(:expires_at, nil)
+      sliding_expires_at = cache_value&.fetch(:sliding_expires_at, nil)
+      # start fetch new token before actual expires time
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled
+        now = Time.zone.now.to_f
+        if sliding_expires_at && now >= sliding_expires_at
+          # if passed sliding expiration time(exp-3*prefetch_ttl)
+          # but not 2*prefetch_ttl from hard expiration time
+          if now > (sliding_expires_at + prefetch_ttl)
+            cache_value = retrieve_token
+          else
+            # extend sliding expiration time
+            cache_value[:sliding_expires_at] = sliding_expires_at + prefetch_ttl
+            save_token(cache_value, expires_at)
```

Comment:
During the initial rollout, we will need to be capable of both writing and reading the token in the previous format. This is because the servers with the previous code will be reading the same token, but will lack the code to parse and use it. After the servers have the new code, we can then update a flag so that they start using the new token format.

Reply:
@dawei-nava Per our discussion, this code still needs to be updated to permit both the previous and new servers to correctly read the token during the rollout. The rollout will not include planned downtime.